File: code/com/leetcode/solutions/NO_0017_电话号码的字母组合/Solution.cs
using System;

public class Solution
{

    /*
    * 思路：通过映射map的到所有的字母组合的List，然后对所有组合中的字符进行组合，同时存入List中，输出即可
    * 其他：
   */

    public IList<string> LetterCombinations(string digits)
    {
        Dictionary<string, string> func = new Dictionary<string, string>();
        func.Add("2", "abc");
        func.Add("3", "def");
        func.Add("4", "ghi");
        func.Add("5", "jkl");
        func.Add("6", "mno");
        func.Add("7", "pqrs");
        func.Add("8", "tuv");
        func.Add("9", "wxyz");
        IList<string> res = new List<string>();

        List<string> rt = new List<string>();
        digits = digits.Replace("1", "");
        foreach (Char v in digits)
        {
            rt.Add(func[v.ToString()]);
        }
        int index = 0;

        for (int i = 1; i < rt.Count; i++)
        {
            if (index == rt.Count - 1) break;
            string source = rt[index];
            foreach (Char item in source)
            {
                for (int k = index; i < rt.Count; k++)
                {
                    foreach (Char vv in rt[k])
                    {
                        res.Add((vv + item).ToString());
                    }
                }
            }
            index++;

        }
        return res;
    }

}

File: code/com/leetcode/solutions/NO_0017_电话号码的字母组合/Solution.java
package com.leetcode.solutions.NO_0017_电话号码的字母组合;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by gang.xu01@hand-china.com on 2020/7/27
 *
 * @author gang.xu01@hand-china.com
 */
public class Solution {

    private Map<String, String> phone = new HashMap<String, String>() {{
        put("2", "abc");
        put("3", "def");
        put("4", "ghi");
        put("5", "jkl");
        put("6", "mno");
        put("7", "pqrs");
        put("8", "tuv");
        put("9", "wxyz");
    }};

    private List<String> output = new ArrayList<>();

    private void deal(String combination, String next) {
        if (next.length() == 0) {
            // 最后一次结束
            output.add(combination);
        } else {
            // 继续组合
            String c = phone.get(next.substring(0, 1));
            for (int i = 0; i < c.length(); i++) {
                this.deal(combination + c.substring(i, i + 1), next.substring(1));
            }
        }
    }

    public List<String> letterCombinations(String digits) {
        if (digits.length() > 0) {
            this.deal("", digits);
        }
        return output;
    }
}
